controllers/storagecluster/cephfilesystem.go

```diff
@@ -19,59 +20,73 @@ type ocsCephFilesystems struct{}
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      generateNameForCephFilesystem(initData),
-			Namespace: initData.Namespace,
+			Name:      generateNameForCephFilesystem(initStorageCluster),
+			Namespace: initStorageCluster.Namespace,
 		},
 		Spec: cephv1.FilesystemSpec{
 			MetadataPool: cephv1.PoolSpec{
-				Replicated:    generateCephReplicatedSpec(initData, "metadata"),
-				FailureDomain: initData.Status.FailureDomain,
+				Replicated:    generateCephReplicatedSpec(initStorageCluster, "metadata"),
+				FailureDomain: initStorageCluster.Status.FailureDomain,
 			},
 			MetadataServer: cephv1.MetadataServerSpec{
 				ActiveCount:   1,
 				ActiveStandby: true,
-				Placement:     getPlacement(initData, "mds"),
-				Resources:     defaults.GetDaemonResources("mds", initData.Spec.Resources),
+				Placement:     getPlacement(initStorageCluster, "mds"),
+				Resources:     defaults.GetDaemonResources("mds", initStorageCluster.Spec.Resources),
 				// set PriorityClassName for the MDS pods
 				PriorityClassName: openshiftUserCritical,
 			},
 		},
 	}
 
-	if initData.Spec.StorageProfiles == nil {
-		// standalone deployment will not have storageProfile, we need to
-		// define default dataPool, if storageProfile is set this will be
-		// overridden.
-		ret.Spec.DataPools = []cephv1.NamedPoolSpec{
-			{
-				PoolSpec: cephv1.PoolSpec{
-					DeviceClass:   generateDeviceClass(initData),
-					Replicated:    generateCephReplicatedSpec(initData, "data"),
-					FailureDomain: initData.Status.FailureDomain,
+	// Load all StorageProfile objects in the StorageCluster's namespace
+	storageProfiles := &ocsv1.StorageProfileList{}
+	spErr := r.Client.List(r.ctx, storageProfiles, &client.ListOptions{Namespace: initStorageCluster.GetNamespace()})
+	if spErr != nil {
+		r.Log.Error(spErr, fmt.Sprint(ocsv1.StorageProfileListError))
+		return nil, fmt.Errorf("%s", ocsv1.StorageProfileListError)
+	}
+
+	if storageProfiles == nil {
+		// not in provider mode
+		if !initStorageCluster.Spec.AllowRemoteStorageConsumers {
+			// standalone deployment that isn't in provider cluster will not
+			// have storageProfile, we need to define default dataPool, if
+			// storageProfile is set this will be overridden.
+			ret.Spec.DataPools = []cephv1.NamedPoolSpec{
+				{
+					PoolSpec: cephv1.PoolSpec{
+						DeviceClass:   generateDeviceClass(initStorageCluster),
+						Replicated:    generateCephReplicatedSpec(initStorageCluster, "data"),
+						FailureDomain: initStorageCluster.Status.FailureDomain,
+					},
 				},
-			},
+			}
 		}
 	} else {
 		// set deviceClass and parameters from storageProfile
-		for i := range initData.Spec.StorageProfiles {
-			deviceClass := initData.Spec.StorageProfiles[i].DeviceClass
-			parameters := initData.Spec.StorageProfiles[i].SharedFilesystemConfiguration.Parameters
+		for i := range storageProfiles.Items {
+			storageProfile := storageProfiles.Items[i]
+			storageProfile.Status.Phase = ocsv1.StorageProfilePhaseInit
```

Comment:
This line is meaningless, as we never persist in the change. If the intent was to mark the fact that we started the reconciliation of a storage profile then we need to persist that change by updating the storage profile CR in the k8s (etcd) database.

Regardless, our reconciliation process is not one that is stretched across multiple reconcile loops (because we always either succeed or fail), so an initializing phase does not make sens as it is guaranteed to be overwritten by the time we exist the reconciliation code. 

I would suggest to remove this phase from the code and from the API (types)